shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +36,97 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
+                builder.Append(_builtInTypeNames[type]);
             }
             else
             {
-                sb.Append(fullName ? t.FullName : t.Name);
+                builder.Append(fullName ? type.FullName : type.Name);
             }
         }
 
-        private static void ProcessNestedGenericTypes(Type t, StringBuilder sb, bool fullName)
+        private static void ProcessGenericType(StringBuilder builder, Type type, bool fullName)
         {
-            var genericFullName = t.GetGenericTypeDefinition().FullName;
-            var genericSimpleName = t.GetGenericTypeDefinition().Name;
-            var parts = genericFullName.Split('+');
-            var genericArguments = t.GetTypeInfo().GenericTypeArguments;
-            var index = 0;
-            var totalParts = parts.Length;
-            if (totalParts == 1)
+            var genericArguments = type.GetGenericArguments();
+            ProcessNestedGenericType(builder, type, genericArguments, 0, fullName);
+        }
+
+        private static int ProcessNestedGenericType(StringBuilder builder, Type type, Type[] genericArgs, int includedArgsCount, bool fullName)
+        {
+            if (fullName)
+            {
+                if (type.IsNested)
+                {
+                    includedArgsCount = ProcessNestedGenericType(builder, type.DeclaringType, genericArgs, includedArgsCount, fullName);
+                    builder.Append('+');
+                }
+                else
+                {
+                    builder.Append(type.Namespace);
+                    builder.Append('.');
+                }
+            }
+            else
             {
-                var part = parts[0];
-                var num = part.IndexOf('`');
-                if (num == -1) return;
+                // skip generic arguments that belongs to nesting types
+                var nestingType = type.DeclaringType;
+                while (nestingType != null)
+                {
+                    includedArgsCount += nestingType.GetGenericArguments().Length;
+                    nestingType = nestingType.DeclaringType;
+                }
+            }
 
-                var name = part.Substring(0, num);
-                var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                sb.Append(fullName ? name : genericSimpleName.Substring(0, genericSimpleName.IndexOf('`')));
-                AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                return;
+            AppendNameOfGenericType(builder, type);
+
+            var numberOfArgs = type.GetGenericArguments().Length;
+            if (numberOfArgs == 0 || includedArgsCount >= genericArgs.Length)
+            {
+                return includedArgsCount;
             }
-            for (var i = 0; i < totalParts; i++)
+
+            var limit = Math.Min(includedArgsCount + numberOfArgs, genericArgs.Length);
```

Comment:
There should not be Math.Min. If there are generic args on this type then we have to traverse from includedArgsCount to number of args to append them.

In theory, IncludedArgsCount will reach it max which is length of genericArgs in the last generic type it has to put. There should not be any mismatch arising.

Reply:
Situation here similar to described in my previous comment.
For type nested inside generic type like `Outer<int>.E<int>` method `GetGenericArguments()` will return two elements (`int` and `int`) and `includedArgsCount + numberOfArgs` will be equal to 3 (`includedArgsCount ` equal 1, becouse one argument was printed for `Outer` class). 

So with the length of  `genericArgs` array equals 2 we will get `IndexOutOfBounds` exception without that `Math.Min`.

Reply:
So it is just question of incorrect math. If we are looking at generic args in nested type then it also includes from declaring type which needs to be subtracted. Using `Math.Min` here for that is not acceptable.